### PR TITLE
Add rule S6319: Using unencrypted SageMaker notebook instances is security-sensitive

### LINKED ISSUE #3439: Add rule S6319: Using unencrypted SageMaker notebook instances is security-sensitive (APPSEC-162)
## What
The rule S6319 detects SageMaker [NotebookInstance](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sagemaker.CfnNotebookInstance.html) that implicitly do not enable encryption at rest in AWS CDK by not defining a KMS key.
This has been previously implemented for Python. See [SONARPY-1130](https://sonarsource.atlassian.net/browse/SONARPY-1130).

## CfnNotebookInstance

- Fully qualified class name: `aws-cdk-lib.aws_sagemaker.CfnNotebookInstance`
- Initializer [`CfnNotebookInstance(scope: Construct, id: string, props: CfnNotebookInstanceProps)`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sagemaker.CfnNotebookInstance.html#initializer)
- Source: `props` (3rd parameter)
- Property: [`kmsKeyId`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sagemaker.CfnNotebookInstance.html#kmskeyid)
- Raise an issue if:
  - `kmsKeyId` property is not set
  - `kmsKeyId` property is set to `undefined`

#### Examples

```javascript
    new CfnNotebookInstance(this, 'noncompliant', {
      instanceType: 'instanceType',
      roleArn: 'roleArn'
    }); // Sensitive: Omitting `kms_key_id` disables encryption of SageMaker notebook instances. Make sure it is safe here.
```

## Code examples

https://github.com/SonarSource/appsec-poc/pull/40

## RSPEC

After implementation, please merge this [RSPEC](https://github.com/SonarSource/rspec/pull/1304).

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**
Fixes #3439

### CODE REVIEW

File: eslint-bridge/tests/linting/eslint/rules/comment-based/aws-sns-unencrypted-topics.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
-const { Topic, CfnTopic } = require('aws-cdk-lib/aws-sns');
-const { Key } = require('aws-cdk-lib/aws-kms');
+import { Topic, CfnTopic } from 'aws-cdk-lib/aws-sns';
+import { Key } from 'aws-cdk-lib/aws-kms';
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/linting/eslint/rules/comment-based/aws-sns-unencrypted-topics.js:2]: Could you revert these changes? I originally defined `aws-sns-unencrypted-topics.js` to validate the `require` resolving and `aws-sns-unencrypted-topics.js` to validate the `import` resolving. In other words, for this first rule, I had two files for the comment-based testing framework.

File: its/sources/file-for-rules/S6319.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+import { CfnNotebookInstance } from "aws-cdk-lib/aws-sagemaker";
+new CfnNotebookInstance(this, "CfnNotebookInstance", { kmsKeyId: undefined });
```
Comment (yassin-kammoun-sonarsource) [inline its/sources/file-for-rules/S6319.js:2]: Could you leave a comment `// Sensitive...` at the line where we expect an issue? It's not used for assertion but still helps to understand where we expect an issue to be raised.

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6319.html [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<p>Amazon SageMaker is a managed machine learning service in a hosted production-ready environment. To train machine learning models, SageMaker
```
Comment (yassin-kammoun-sonarsource) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6319.html:1]: Don't forget to merge the RSPEC.

File: eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts [hunk 4]
```diff
@@ -45,7 +45,13 @@ export const rule: Rule.RuleModule = AwsCdkTemplate(
 
 function checkTopic(key: string) {
   return (expr: estree.NewExpression, ctx: Rule.RuleContext) => {
-    const props = getValueOfExpression(ctx, expr.arguments[2], 'ObjectExpression');
+    const argument = expr.arguments[2];
+    const props = getValueOfExpression(ctx, argument, 'ObjectExpression');
+
+    if (argument?.type === 'Identifier' && !isUndefined(argument) && props === undefined) {
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:51]:
```suggestion
    if (isUnknown(argument, props)) {
```

Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:51]: That's a little bit tricky, and I am not sure we will easily remember what this means. Should we add a local function whose name tells what this check is all about?

**Review by yassin-kammoun-sonarsource — CHANGES_REQUESTED** [review]

File: eslint-bridge/tests/linting/eslint/rules/comment-based/aws-sns-unencrypted-topics.js (hunk 1 above)
Comment (victor-diez-sonarsource) [reply eslint-bridge/tests/linting/eslint/rules/comment-based/aws-sns-unencrypted-topics.js:2]: oopps, sorry! reverting

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

**Pull request merged**